stl/inc/functional

```diff
@@ -1479,6 +1479,83 @@ namespace placeholders {
     _INLINE_VAR constexpr _Ph<20> _20{};
 } // namespace placeholders
 
+#if _HAS_CXX20
+// FUNCTION TEMPLATE _Call_front_binder
+template <size_t... _Ix, class _Cv_FD, class _Cv_tuple_TiD, class... _Unbound>
+constexpr auto _Call_front_binder(index_sequence<_Ix...>, _Cv_FD&& _Obj, _Cv_tuple_TiD&& _Tpl,
+    _Unbound&&... _Unbargs) noexcept(noexcept(_STD invoke(_STD forward<_Cv_FD>(_Obj),
+    _STD get<_Ix>(_STD forward<_Cv_tuple_TiD>(_Tpl))..., _STD forward<_Unbound>(_Unbargs)...)))
+    -> decltype(_STD invoke(_STD forward<_Cv_FD>(_Obj), _STD get<_Ix>(_STD forward<_Cv_tuple_TiD>(_Tpl))...,
+        _STD forward<_Unbound>(_Unbargs)...)) {
+    return _STD invoke(_STD forward<_Cv_FD>(_Obj), _STD get<_Ix>(_STD forward<_Cv_tuple_TiD>(_Tpl))...,
+        _STD forward<_Unbound>(_Unbargs)...);
+}
+
+// CLASS TEMPLATE _Front_binder
+template <class _Fx, class... _Types>
+class _Front_binder { // wrap bound callable object and arguments
+private:
+    static_assert(is_constructible_v<decay_t<_Fx>, _Fx>,
+        "std::bind_front() requires the decayed callable to be constructible from an undecayed callable");
+    static_assert(is_move_constructible_v<decay_t<_Fx>>,
+        "std::bind_front() requires the decayed callable to be move constructible");
+    static_assert(conjunction_v<is_constructible<decay_t<_Types>, _Types>...>,
+        "std::bind_front() requires the decayed bound arguments to be constructible from undecayed bound arguments");
+    static_assert(conjunction_v<is_move_constructible<decay_t<_Types>>...>,
+        "std::bind_front() requires the decayed bound arguments to be move constructible");
+
+    using _Seq    = index_sequence_for<_Types...>;
+    using _First  = decay_t<_Fx>;
+    using _Second = tuple<decay_t<_Types>...>;
+
+    _Compressed_pair<_First, _Second> _Mypair;

```

Comment:
I disagree. `tuple` doesn't perform compression in our current ABI, while `_Compressed_pair` will compress away an empty callable object. This can make the difference between dynamically allocating or not, if the `bind_front()` object is stored in a `std::function`. For this reason, I believe that Nathan's current implementation is ideal.

If you would like to open a `vNext performance enhancement` issue to track using `tuple` in all of the binders, please go ahead.

Reply:
All good points, and I'll throw in "extending the `tuple` is likely more costly in throughput terms than instantiating `_Compressed_pair` given our recursive `tuple` implementation." Filed as #197.

I'll narrow this comment to simply "`_First` and `_Second` are used only once each; we could as easily use `_Compressed_pair<decay_t<_Fx>, tuple<decay_t<_Types>...>>` directly as the type of `_Mypair`."

Reply:
The primary reason I used the `_First` and `_Seconds` aliases was to match the previous implementation of `bind()`'s style.
https://github.com/microsoft/STL/blob/1e4f423120b1b3de409655bdbcc0492121940afb/stl/inc/functional#L1421-L1428

Reply:
I like the consistency there (but of course, I wrote the current implementation of `bind()`).

@CaseyCarter do you want both to be changed? I am very weakly in favor of the status quo.

Reply:
I'm as weakly in favor of changing both - these type aliases cost dozens of nanoseconds of compile time! I suppose @NathanSWard can decide to either make the change or leave the PR as-is and mark this comment resolved either way.